io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -1079,130 +1090,290 @@ traceMany ((time, tid, event):ts) trace =
 -- Executing STM Transactions
 --
 
-data TVar s a = TVar !TVarId
-                     !(STRef s a)          -- current value
-                     !(STRef s (Maybe a))  -- saved revert value
-                     !(STRef s [ThreadId]) -- threads blocked on read
+data TVar s a = TVar {
+
+       -- | The identifier of this var.
+       --
+       tvarId :: !TVarId,
+
+       -- | The var's current value
+       --
+       tvarCurrent :: !(STRef s a),
+
+       -- | A stack of undo values. This is only used while executing a
+       -- transaction.
+       --
+       tvarUndo    :: !(STRef s [a]),
+
+       -- | Thread Ids of threads blocked on a read of this var. It is
+       -- represented in reverse order of thread wakeup, without duplicates.
+       --
+       -- To avoid duplicates efficiently, the operations rely on a copy of the
+       -- thread Ids represented as a set.
+       --
+       tvarBlocked :: !(STRef s ([ThreadId], Set ThreadId))
+     }
 
 data StmTxResult s a =
-       StmTxComitted a
-                     [TVarId]
-                     [(ThreadId, [TVarId])] -- wake up
-                     TVarId                 -- updated TVarId name supply
+       -- | A comitted transaction reports the vars that were written (in order
+       -- of first write) so that the scheduler can unblock other threads that
+       -- were blocked in STM transactions that read any of these vars.
+       --
+       -- It also includes the updated TVarId name supply.
+       --
+       StmTxComitted a [SomeTVar s] TVarId -- updated TVarId name supply
+
+       -- | A blocked transaction reports the vars that were read so that the
+       -- scheduler can block the thread on those vars.
+       --
+     | StmTxBlocked  [SomeTVar s]
      | StmTxAborted  SomeException
-     | StmTxBlocked  [TVarId]               -- blocked on
 
 data SomeTVar s where
   SomeTVar :: !(TVar s a) -> SomeTVar s
 
-execAtomically :: ThreadId
-               -> TVarId
+data StmStack s b a where
+  -- | Executing in the context of a top level 'atomically'.
+  AtomicallyFrame  :: StmStack s a a
+  
+  -- | Executing in the context of the /left/ hand side of an 'orElse'
+  OrElseLeftFrame  :: StmA s a                -- orElse right alternative
+                   -> (a -> StmA s b)         -- subsequent continuation
+                   -> Map TVarId (SomeTVar s) -- saved written vars set
+                   -> [SomeTVar s]            -- saved written vars list
```

Comment:
Do we not want to use haddock annotations here?

Reply:
Unfortunately, haddoc annotations cannot be used this way over `GADT`s.